Reorder xp tracker bars

Shingyx opened the pull request:
Implement a custom MouseAdapter which watches for mouse drag events to move child components within a JPanel. Include a workaround for child popup menus, since setComponentPopupMenu consumes the required MouseEvents. Also remove redundant line in `XpInfoBox#reset`.

Resolves https://github.com/runelite/runelite/issues/3102 and https://github.com/runelite/runelite/issues/2029. There could be others too, I didn't look that hard.

The mouse adapter should be generic enough so that it could be applied to some of the other plugins in the future.

Shingyx commented:
I think I've caught most of the edge cases. Would be interested in your thoughts now. Honestly, I think the best way to review this is pull it down locally and experiment with it. Swing doesn't make drag and drop easy, unless I completely missed something in my Google searches.

Gif demonstrating the feature:
![drag](https://user-images.githubusercontent.com/11037113/42151783-2776d9f2-7e32-11e8-84e6-9e66802477b5.gif)

Shingyx commented on runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java:
```diff
@@ -69,13 +69,14 @@
 	private final JLabel actionsLeft = new JLabel();
 
-	XpInfoBox(XpTrackerPlugin xpTrackerPlugin, Client client, JPanel panel, Skill skill, SkillIconManager iconManager) throws IOException
+	XpInfoBox(XpTrackerPlugin xpTrackerPlugin, Client client, XpInfoBoxPanel panel, Skill skill, SkillIconManager iconManager) throws IOException
```
Any reason why this constructor declares `throws IOException`? I can't see anything that could throw one.

Shingyx commented:
I don't know of any more issues from testing on Windows. However, I tried this out on Ubuntu (and Kubuntu) and it works great in most cases, except for one issue I noticed where it can get into a weird state when trying to drag an info panel while a pop up menu for another info panel is open, where the dragging panel gets "stuck" when the mouse is released. This is because the `MouseReleased` event doesn't seem to get received when this happens. The overall UI recoverable by clicking somewhere else in the XP plugin panel.

On that note, should the reordering be configurable? I don't know how I'd workaround this issue...

Shingyx commented on runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java:
```diff
@@ -69,13 +70,14 @@
 	private final JLabel actionsLeft = new JLabel();
 
-	XpInfoBox(XpTrackerPlugin xpTrackerPlugin, Client client, JPanel panel, Skill skill, SkillIconManager iconManager) throws IOException
+	XpInfoBox(XpTrackerPlugin xpTrackerPlugin, Client client, JComponent panel, Skill skill, SkillIconManager iconManager) throws IOException
```
Does `throws IOException` need to be here?

Shingyx commented:
Rebased and squashed.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/xptracker/DragAndDropReorderPane.java:
```diff
@@ -0,0 +1,302 @@
+ * for drag and drop reordering. Instead, implement the PopupMenuOwner interface in the children components to enable
+ * both child popup menus and drag and drop reordering.
+ */
+public class DragAndDropReorderPane extends JLayeredPane
```
If this is going to be a public class, this should not be introduced as a plugin class. This should be added with the other UI elements in runelite/runelite-client/src/main/java/net/runelite/client/ui/.

Shingyx replied:
Done. I moved both DragAndDropReorderPane and PopupMenuOwner. `net.runelite.client.ui.components` seemed like the most appropriate package.

Shingyx commented:
I've rebased again. Is there anything else I can do to help move this PR forwards? It's been almost a year now.

Personally, the XP tracker is my most used feature of RuneLite and I've always found it annoying when the only way to reorder the tracker bars is to reset the ones which I want to move to the bottom, especially since this also resets the XP rate calculation.

I understand that a lot of code was added, but Swing's mouse event handling made adding drag-and-drop while maintaining popup menus difficult.

Nightfirecat commented:
I'll try to give this a review this weekend. I appreciate you've kept this PR up to date, and I'm sure we can find some time to properly review this and get it merged.

Shingyx commented:
That's awesome to hear! Please let me know if there are any issues or if anything needs clarification.

Thanks!

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/ui/components/DragAndDropReorderPane.java:
```diff
@@ -0,0 +1,250 @@
+ */
+public class DragAndDropReorderPane extends JLayeredPane
+{
+	private Point dragStartPoint = null;
```
Non-primitive fields are initialized to `null` by default; there is no need to explicitly set them.
```suggestion
	private Point dragStartPoint;
```

Shingyx replied:
Removed

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/ui/components/DragAndDropReorderPane.java:
```diff
@@ -0,0 +1,250 @@
+/**
+ * Pane which allows reordering its components via drag and drop.
+ * <p>
+ * For child components' popup menus, try avoid using setComponentPopupMenu as this consumes the MouseEvents required
```
This doesn't feel like entirely the right place for this comment. This should probably be moved to the PopupMenuOwner interface so it can be more easily found.

Shingyx replied:
I've split the comment out a bit. Is it better now?

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/plugins/xptracker/XpInfoBox.java:
```diff
@@ -197,7 +199,6 @@
 	void reset()
 	{
 		canvasItem.setText(ADD_STATE);
-		container.remove(statsPanel);
```
Why is this line removed?

Shingyx replied:
`statsPanel` was moved to be inside `headerPanel` at some point, so this line had no effect.

Nightfirecat commented on runelite-client/src/main/java/net/runelite/client/ui/components/DragAndDropReorderPane.java:
```diff
@@ -0,0 +1,250 @@
+		y = Math.min(y, getHeight() - draggingComponent.getHeight());
+
+		Point location = new Point(0, y);
+		draggingComponent.setLocation(location);
```
Remove the `location` variable above and set the location with the new Point directly.
```suggestion
		draggingComponent.setLocation(new Point(0, y));
```

Shingyx replied:
Done

Nightfirecat commented:
Also, it looks like the second commit can be squashed into the first? This change looks like it can be contained in one commit.

Shingyx commented:
> 
> 
> Also, it looks like the second commit can be squashed into the first? This change looks like it can be contained in one commit.

I've squashed and rebased

raiyni commented:
You don't need to comment on every request. Just click the mark as resolved button.

Nightfirecat commented:
Code LGTM now, but I'll freely admit that I'm not well-versed in Swing, so some other contributors should probably weigh in on that. Thanks for responding so quickly to feedback. :+1:

raiyni commented:
1. If this doesn't save the order of the panels, what is the point?
2. This maybe could be used to resolve #783 if implemented correctly

Shingyx commented:
Personally, I don't think that saving the order of the panels between sessions is necessary for this to be useful. One example use case: I want to keep all XP trackers active for the whole session, so that I can see how much XP I gained in each skill at the end, but at the same time I want the skill I'm actively working on to be at the top.

https://github.com/runelite/runelite/issues/783 looks out of scope to me. I think the component I added could be used for that though.

Shingyx commented:
I've rewritten basically half of this PR for the following:
* Removed the pop-up menu hacks since I've found a relatively clean way to forward mouse drag events to parent components. Replaced with a `MouseDragEventForwarder` instead.
* Improved how the XP info box follows the mouse, avoiding the "jump" when the dragging starts.

Please let me know if there's anything else I can do to move this forward. The linked issue has been duplicated several times now, suggesting that some people are interested in this feature.

Shingyx commented:
Rebased and updated gif.

deathbeam commented on runelite-client/src/main/java/net/runelite/client/ui/components/DragAndDropReorderPane.java:
```diff
@@ -0,0 +1,203 @@
+				Point point = e.getPoint();
+				if (!contains(point))
+				{
+					finishDragging();
```
Imo this should not stop dragging when OOB. Mainly when OOB means the size of parent container what is size of the actual listed xp trackers, so the dragging just stops there and it is pretty annoying.

Shingyx replied:
Yeah that makes sense. Updated.
![sample4](https://user-images.githubusercontent.com/11037113/83957601-3f269580-a8bd-11ea-9c67-1fe28b5ddb5b.gif)

abextm commented:
Why doesn't this use swing's drag and drop stuff?

Shingyx commented:
There were limitations with all of the native Swing implementations I could find. None of them were designed for **reordering** components (like the `XpInfoBox`) inside a `JComponent`/`JPanel` with drag and drop. The best examples I found involved writing a custom implementation similar to what I have done here.

There is a native drag and drop implementation for [JList](https://docs.oracle.com/javase/7/docs/api/javax/swing/JList.html#setDragEnabled(boolean)), but rewriting the XP panel to be a `JList` holding `XpInfoBox` components doesn't sound like a good idea. The other [Swing drag and drop stuff](https://docs.oracle.com/javase/tutorial/uiswing/dnd/intro.html) I could find mainly involved dragging and dropping **data** in the application.

So from what I could find, [JLayeredPane](https://docs.oracle.com/javase/7/docs/api/javax/swing/JLayeredPane.html) is the correct choice here. It defines a `DRAG_LAYER` layer for rendering components being dragged around. I admit I'm not the best at Swing, so there could be _other_ ways to implement this, but IMO this already works really well as-is.

Shingyx commented on runelite-client/src/main/java/net/runelite/client/ui/components/DragAndDropReorderPane.java:
```diff
@@ -0,0 +1,199 @@
+
+import javax.swing.BoxLayout;
+import javax.swing.JLayeredPane;
+import javax.swing.SwingUtilities;
```
I must've wrote that before configuring IntelliJ properly 😅
